<!DOCTYPE html>
<html lang="en"><head>
<script src="02_data_processing_files/libs/clipboard/clipboard.min.js"></script>
<script src="02_data_processing_files/libs/quarto-html/tabby.min.js"></script>
<script src="02_data_processing_files/libs/quarto-html/popper.min.js"></script>
<script src="02_data_processing_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="02_data_processing_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="02_data_processing_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="02_data_processing_files/libs/quarto-html/quarto-syntax-highlighting-226bd0f977fa82dfae4534cac220d79a.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.23">

  <meta name="author" content="Kerry Cobb">
  <title>Single Cell Raw Data Processing</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="02_data_processing_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="02_data_processing_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="02_data_processing_files/libs/revealjs/dist/theme/quarto-250687541b6fd0303e6fa9a425de9713.css">
  <link href="02_data_processing_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="02_data_processing_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="02_data_processing_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="02_data_processing_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link href="custom.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Single Cell Raw Data Processing</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Kerry Cobb 
</div>
</div>
</div>

</section>
<section id="goals" class="slide level2">
<h2>Goals</h2>
<ul>
<li>Understand the steps for processing raw single cell RNA-seq data</li>
<li>See how to run software for each processing step</li>
<li>See scripts for running each piece of software on HPC</li>
</ul>
</section>
<section id="dataset" class="slide level2 smaller">
<h2>Dataset</h2>
<ul>
<li>Childhood acute lymphoblastic leukemia</li>
<li>Most common childood cancer</li>
<li>Precursor B-cell acute lymphoblastic leukemia (B-ALL)
<ul>
<li>~85% of cases</li>
<li>Good prognosis</li>
</ul></li>
<li>T-cell acute lymphoblastic leukemia (T-ALL)
<ul>
<li>~15% of cases</li>
</ul></li>
<li>Other patients are refractory or will relapse and succumb</li>
<li>Compare 4 B-ALL, 2 T-ALL, and 3 healthy controls</li>
<li>Cells are Bone Marrow Mononuclear cells (BMMCs)</li>
<li>Each cell sequenced on a different lane of Illumina
<ul>
<li>What’s wrong with that?</li>
</ul></li>
</ul>
</section>
<section id="workflow" class="slide level2">
<h2>Workflow</h2>
<ol type="1">
<li>Data download</li>
<li>Quality control</li>
<li>Mapping &amp; alignment</li>
<li>Feature counting</li>
</ol>
</section>
<section id="logging-in-to-the-hpc" class="slide level2">
<h2>Logging in to the HPC</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href=""></a><span class="fu">ssh</span> genomeuser<span class="op">&lt;</span><span class="co">#&gt;@xanadu-submit-ext.cam.uchc.edu</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="git-repository" class="slide level2">
<h2>Git Repository</h2>
<ul>
<li><p>All of the code and data for this workshop is available on GitHub.</p></li>
<li><p>You can clone the repository to your home directory on the HPC using the following command:</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href=""></a><span class="fu">git</span> clone https://github.com/CBC-UCONN/Single-Cell-Transcriptomics.git</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="symlink-to-existing-output-data" class="slide level2">
<h2>Symlink to existing output data</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href=""></a><span class="bu">cd</span> ~/Single-Cell-Transcriptomics/leukemia/</span>
<span id="cb3-2"><a href=""></a><span class="fu">ln</span> <span class="at">-s</span> /core/cbc/tutorials/workshopdirs/Single-Cell-Transcriptomics/leukemia/results/</span>
<span id="cb3-3"><a href=""></a><span class="bu">cd</span> ~/Single-Cell-Transcriptomics/resources/Homo_sapiens.GRCh38 </span>
<span id="cb3-4"><a href=""></a><span class="fu">ln</span> <span class="at">-s</span> /core/cbc/tutorials/workshopdirs/Single-Cell-Transcriptomics/resources/Homo_sapiens.GRCh38/Homo_sapiens.GRCh38.114.chr.gtf</span>
<span id="cb3-5"><a href=""></a><span class="fu">ln</span> <span class="at">-s</span> /core/cbc/tutorials/workshopdirs/Single-Cell-Transcriptomics/resources/Homo_sapiens.GRCh38/Homo_sapiens.GRCh38.dna.primary_assembly.fa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li>You would skip this if doing these analyses on your own</li>
<li>Some of these steps take a long time to run, so we have provided the output data for you to use in the workshop.</li>
<li><strong>Please don’t copy these data on the cluster!</strong></li>
</ul>
</section>
<section id="reference-genome" class="slide level2">
<h2>Reference Genome</h2>
<ul>
<li>Required for mapping reads</li>
<li>NCBI &amp; Ensemble are good source</li>
<li>For examples see:
<ul>
<li><code>resources/Homo_sapiens.GRCh38/download.sh</code></li>
<li><code>resources/GRCh38.p14/GRCh38_download.sh</code></li>
</ul></li>
</ul>
</section>
<section id="fasta-format" class="slide level2">
<h2>FASTA Format</h2>
<ul>
<li>Common format for storing nucleotide sequences</li>
<li>Typically used for reference genomes</li>
<li>Each sequence is represented by two lines:
<ol type="1">
<li>Identifier line (starts with ‘&gt;’)</li>
<li>Sequence line (the actual nucleotide sequence)</li>
</ol></li>
<li>Example:</li>
</ul>
<pre><code>&gt;SEQ_ID
GATTTGGGGTTTAAAGGGTGACCTGGTAGG</code></pre>
</section>
<section id="data-download" class="slide level2">
<h2>Data Download</h2>
<ul>
<li>NCBI Sequence Read Archive (SRA)</li>
<li>https://www.ncbi.nlm.nih.gov/bioproject/PRJNA548203</li>
<li>Use SRA Toolkit to download</li>
<li><code>include-technical</code>
<ul>
<li>Necessary to obtain Cell and UMI barcodes</li>
</ul></li>
<li>See: <code>leukemia/scripts/01_download.sh</code></li>
</ul>
</section>
<section id="fastq-format" class="slide level2 smaller">
<h2>FASTQ Format</h2>
<ul>
<li>Typical raw data format for sequencing reads</li>
<li>Contains sequence data and quality scores</li>
<li>Typically separated into two or more files
<ul>
<li>One for each end of a read, if paired</li>
<li>Technical reads such as barcodes or UMIs</li>
</ul></li>
<li>Each read is represented by four lines:
<ol type="1">
<li>Header line (starts with ‘@’)</li>
<li>Sequence line (the actual nucleotide sequence)</li>
<li>Separator line (starts with ‘+’)</li>
<li>Quality score line (ASCII-encoded quality scores for each nucleotide in the sequence)</li>
</ol></li>
<li>Example:</li>
</ul>
<pre><code>@SEQ_ID
GATTTGGGGTTTAAAGGGTGACCTGGTAGG
+
!''*((((***+))%%%++)(****)())-+**--1</code></pre>
</section>
<section id="fastq-format---quality-scores" class="slide level2 smaller">
<h2>FASTQ Format - Quality Scores</h2>
<ul>
<li>Quality scores are encoded using ASCII characters</li>
<li>Each character represents a 2 digit quality score for the corresponding nucleotide
<ul>
<li>This permits more efficient storage</li>
</ul></li>
<li>The quality score is a measure of the confidence in the accuracy of the base call</li>
<li>The higher the quality score, the more reliable the base call</li>
<li>Calculated as <code>-10 * log10(P)</code> where <code>P</code> is the probability of an incorrect base call</li>
<li>Examples:
<ul>
<li>Q = 10, P = 0.1 (10% chance of error)</li>
<li>Q = 30, P = 0.001 (0.1% chance of error)</li>
</ul></li>
</ul>
</section>
<section id="fastq-format---headers" class="slide level2 smaller">
<h2>FASTQ Format - Headers</h2>
<ul>
<li>The header line contains metadata about the read</li>
<li>Illumina headers typicallly look like this:</li>
</ul>
<pre><code>@HWUSI-EAS100R:6:73:941:1973#0/1</code></pre>
<p>The components of the header are:</p>
<ul>
<li>HWUSI-EAS100R: the unique instrument name</li>
<li>6: flowcell lane</li>
<li>73: tile number within the flowcell lane</li>
<li>941: ‘x’-coordinate of the cluster within the tile</li>
<li>1973: ‘y’-coordinate of the cluster within the tile</li>
<li>#0: index number for a multiplexed sample (0 for no indexing)</li>
<li>/1: the member of a pair, /1 or /2</li>
</ul>
</section>
<section id="annotations" class="slide level2 smaller">
<h2>Annotations</h2>
<ul>
<li>Annotations provide additional information about the sequences</li>
<li>They can include gene names, descriptions, and other metadata</li>
<li>Annotations are often stored in separate files, such as GTF or GFF files</li>
<li>GTF (Gene Transfer Format) and GFF (General Feature Format) are common formats</li>
<li>Annotation files contain information about gene structures, such as exons, introns, and transcripts</li>
<li>GTF &amp; GFF files are similar but GFF can include more complex features and relationships</li>
<li>GFF is preferred</li>
</ul>
</section>
<section id="annotations-1" class="slide level2 smaller">
<h2>Annotations</h2>
<ul>
<li>Can be downloaded from Ensembl or NCBI</li>
<li>10X Genomics provides a custom annotation file for human and mouse</li>
</ul>
</section>
<section id="quality-control" class="slide level2">
<h2>Quality Control</h2>
<ul>
<li>Quality of reads</li>
<li>Quantify of reads</li>
<li>Adapters Contamination</li>
</ul>
</section>
<section id="adapter-contamination" class="slide level2">
<h2>Adapter Contamination</h2>
<p>Can occur if insert size is too short</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/illumina_sequencing-illumina.jpg" class="nostretch" width="500"></p>
<figcaption>Illumina Inc.</figcaption>
</figure>
</div>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/adapter_contamination-ecseq.png" class="nostretch"></p>
<figcaption>ECSEQ</figcaption>
</figure>
</div>
</section>
<section id="fastqc" class="slide level2">
<h2>Fastqc</h2>
<ul>
<li>Evaluates several metrics for a FASTQ file</li>
<li>Provides a summary of the quality of the reads</li>
<li>Generates an HTML report with visualizations and statistics</li>
<li>Can be run with the command:</li>
</ul>
</section>
<section id="multiqc" class="slide level2">
<h2>Multiqc</h2>
<ul>
<li>Consolidates multiple reports from FastQC and other tools</li>
<li>Generates a single HTML report with visualizations and statistics</li>
<li>Useful for comparing multiple samples or conditions</li>
</ul>
</section>
<section id="mapping-alignment" class="slide level2 smaller">
<h2>Mapping &amp; Alignment</h2>
<ul>
<li>Determination of which part of the genome a read comes from</li>
<li>Can be done with or without alignment</li>
<li>Popular tools include:
<ul>
<li>Cell Ranger (commercial software from 10x Genomics) [Zheng et al., 2017]</li>
<li>zUMIs [Parekh et al., 2018]</li>
<li>alevin [Srivastava et al., 2019]</li>
<li>RainDrop [Niebler et al., 2020]</li>
<li>kallisto|bustools [Melsted et al., 2021]</li>
<li>STARsolo [Kaminow et al., 2021]</li>
<li>alevin-fry [He et al., 2022]</li>
</ul></li>
</ul>
</section>
<section id="mapping-alignment-1" class="slide level2">
<h2>Mapping &amp; Alignment</h2>
<ul>
<li>Methods can be broadly categorized into three types:
<ol type="1">
<li>Spliced alignment</li>
<li>Contiguous alignment</li>
<li>Lightweight mapping (pseudoalignment)</li>
</ol></li>
<li>Alignment-based methods provide easily interpreted mapping scores</li>
<li>Lightweight methods are faster</li>
</ul>
</section>
<section id="spliced-alignment" class="slide level2">
<h2>Spliced Alignment</h2>
<ul>
<li>Reads can align across multiple distinct segments of a reference <img data-src="img/alignment.png"></li>
</ul>
</section>
<section id="contiguous-alignment" class="slide level2">
<h2>Contiguous Alignment</h2>
<ul>
<li>Reads require a continuous segment of the reference</li>
<li>Large gaps such as introns are not allowed</li>
<li>Faster</li>
<li>Requires annotated transcript sequences as reference
<ul>
<li>Not suitable for single nucleus RNA-seq data</li>
</ul></li>
<li>Can be used with augmented transcriptome
<ul>
<li>Full length unspliced transcripts or excised intronic sequences</li>
</ul></li>
</ul>
</section>
<section id="starsolo" class="slide level2">
<h2>STARsolo</h2>
<ul>
<li>Much faster alternative to Cell Ranger</li>
<li>Produces near identical output to Cell Ranger</li>
<li>Cell Ranger uses STAR for alignment</li>
<li>STAR can be used for other types of single cell RNA-seq data
<ul>
<li>No need to learn a new tool or make large modifications to the workflow</li>
</ul></li>
<li>Requires an index for the reference genome}</li>
</ul>
</section>
<section id="starsolo-indexing" class="slide level2">
<h2>STARsolo Indexing</h2>
<ul>
<li><code>-sjdbOverhang</code>
<ul>
<li>Length of the longest read minus 1</li>
<li>For example, if the longest read is 150 bp, then <code>-sjdbOverhang</code> should be set to 149</li>
</ul></li>
</ul>
</section>
<section id="starsolo-args" class="slide level2 smaller">
<h2>STARsolo Args</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><code>soloCBwhitelist</code>
<ul>
<li>The allowed cell barcodes</li>
<li>These are packged with cell ranger</li>
</ul></li>
<li><code>soloUMIlen</code></li>
<li><code>soloType</code>
<ul>
<li>Type of single-cell RNA-seq</li>
<li>Probaly will want to use CB_UMI_Simple
<ul>
<li>One UMI and one cell barcode of fixed length</li>
</ul></li>
</ul></li>
</ul>
</div><div class="column" style="width:50%;">
<ul>
<li><code>soloUMIlen</code>
<ul>
<li>10 for v2</li>
<li>12 for &gt;v3</li>
</ul></li>
<li><code>readFilesIn</code>
<ul>
<li>1st file is cDNA read</li>
<li>2nd file barcode (cell+UMI)</li>
</ul></li>
<li><code>readFilesCommand</code>
<ul>
<li><code>zcat</code> if gzipped</li>
</ul></li>
<li><code>soloOutFormatFeaturesGzipped</code>
<ul>
<li>compress output mtx files, <code>scanpy</code> expects these to be .gz</li>
</ul></li>
</ul>
</div></div>
</section>
<section id="sam-bam-file-formats" class="slide level2 smaller">
<h2>SAM &amp; Bam file formats</h2>
<ul>
<li>Output is a SAM file</li>
<li>SAM (Sequence Alignment/Map) is a text format for storing aligned reads</li>
<li>It contains information about the alignment of each read to the reference genome</li>
<li>SAM files are human-readable, but can be large</li>
<li>They can be compressed to BAM format for storage and efficiency
<ul>
<li>They are compressed and indexed for efficient access</li>
<li>Not human-readable</li>
</ul></li>
<li>SAM/BAM files can be viewed with tools like IGV or samtools</li>
<li>They are used for downstream analyses like feature counting and variant calling</li>
<li>Example SAM file: <code>@HD  VN:1.0  SO:unsorted @SQ SN:chr1 LN:248956422 @RG ID:group1   SM:sample1 @PG ID:program1 PN:STAR VN:2.7.9a read1   0   chr1    100 60  10M1I5M2D3M *   0   0   ACGTACGTAC  IIIIIIIIIIIIIIIIIIII    AS:i:10 NM:i:1</code></li>
</ul>
</section>
<section id="cigar-string" class="slide level2 smaller">
<h2>CIGAR String</h2>
<ul>
<li>CIGAR (Compact Idiosyncratic Gapped Alignment Report) string is a part of the SAM/BAM format</li>
<li>It describes how the read aligns to the reference genome</li>
<li>It consists of a series of operations, each represented by a character and a number</li>
<li>Common operations include:
<ul>
<li>M: match or mismatch</li>
<li>I: insertion</li>
<li>D: deletion</li>
<li>N: skipped region (intron)</li>
<li>S: soft clipping (part of the read is not aligned)</li>
</ul></li>
<li>Example CIGAR string: <code>10M1I5M2D3M</code>
<ul>
<li>This means 10 matches, 1 insertion, 5 matches, 2 deletions, and 3 matches</li>
</ul></li>
</ul>
</section>
<section id="feature-counting" class="slide level2">
<h2>Feature Counting</h2>
<ul>
<li>Quantify abundance of each gene using UMIs</li>
<li>STAR-solo also does feature counting for us</li>
</ul>
</section>
<section id="feature-counts" class="slide level2">
<h2>Feature Counts</h2>

<img data-src="img/matrix.png" class="r-stretch"></section>
<section id="single-cell-analysis" class="slide level2">
<h2>Single Cell Analysis</h2>
<p>Up next!</p>

</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="02_data_processing_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="02_data_processing_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="02_data_processing_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="02_data_processing_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>