<!DOCTYPE html>
<html lang="en"><head>
<script src="06_analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="06_analysis_files/libs/quarto-html/tabby.min.js"></script>
<script src="06_analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="06_analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="06_analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="06_analysis_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="06_analysis_files/libs/quarto-html/quarto-syntax-highlighting-226bd0f977fa82dfae4534cac220d79a.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.23">

  <meta name="author" content="Kerry Cobb">
  <title>Single Cell RNAseq Analysis</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="06_analysis_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="06_analysis_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="06_analysis_files/libs/revealjs/dist/theme/quarto-250687541b6fd0303e6fa9a425de9713.css">
  <link href="06_analysis_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="06_analysis_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="06_analysis_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="06_analysis_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link href="custom.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Single Cell RNAseq Analysis</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Kerry Cobb 
</div>
</div>
</div>

</section>
<section id="steps" class="slide level2">
<h2>Steps</h2>
<ol type="1">
<li>Quality control</li>
<li>Normalization</li>
<li>Feature selection</li>
<li>Dimensionality reduction</li>
<li>Integration</li>
<li>Clustering</li>
<li>Annotation</li>
</ol>
</section>
<section id="steps-1" class="slide level2">
<h2>Steps</h2>
<ol type="1">
<li>Quality control</li>
<li>Normalization</li>
<li>Feature selection</li>
<li>Dimensionality reduction</li>
<li>Integration</li>
<li>Clustering</li>
<li>Annotation</li>
</ol>
<ul>
<li>Not necessarily a linear process</li>
</ul>
</section>
<section id="quality-control" class="slide level2 smaller">
<h2>Quality Control</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><strong>Filter empty droplets</strong></li>
<li>Filter droplets with multiple cells</li>
<li>Remove low quality cells
<ul>
<li>Cells with high mitochondrial gene counts
<ul>
<li>Suggests cell death or damage</li>
</ul></li>
</ul></li>
<li>Filter cells with low gene counts</li>
<li>Filter genes expressed in few cells</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="img/empty_droplet-biorender.png" width="200"></p>
</div></div>
</section>
<section id="quality-control-1" class="slide level2 smaller">
<h2>Quality Control</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><strong>Filter empty droplets</strong></li>
<li>Filter droplets with multiple cells</li>
<li>Remove low quality cells
<ul>
<li>Cells with high mitochondrial gene counts
<ul>
<li>Suggests cell death or damage</li>
</ul></li>
</ul></li>
<li>Filter cells with low gene counts</li>
<li>Filter genes expressed in few cells</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="img/droplet_ambient-biorender.png" width="200"></p>
</div></div>
</section>
<section id="quality-control-2" class="slide level2 smaller">
<h2>Quality Control</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Filter empty droplets</li>
<li><strong>Filter droplets with multiple cells</strong></li>
<li>Remove low quality cells
<ul>
<li>Cells with high mitochondrial gene counts
<ul>
<li>Suggests cell death or damage</li>
</ul></li>
</ul></li>
<li>Filter cells with low gene counts</li>
<li>Filter genes expressed in few cells</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="img/doublet-biorender.png" width="200"></p>
</div></div>
</section>
<section id="quality-control-3" class="slide level2 smaller">
<h2>Quality Control</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Filter empty droplets</li>
<li>Filter droplets with multiple cells</li>
<li>Remove low quality cells
<ul>
<li>Cells with high mitochondrial gene counts
<ul>
<li>Suggests cell death or damage</li>
</ul></li>
</ul></li>
<li>Filter cells with low gene counts</li>
<li>Filter genes expressed in few cells</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img data-src="img/ruptered_cell.png"></p>
</div></div>
</section>
<section id="quality-control-4" class="slide level2 smaller">
<h2>Quality Control</h2>
<ul>
<li>Can set hard filtering thresholds
<ul>
<li>These end up being arbitrary as there is rarely a clear boundary</li>
<li>Time intensive
<ul>
<li>Doesn’t scale well with large datasets</li>
</ul></li>
</ul></li>
<li>Median Absolute Deviation (MAD) based filtering
<ul>
<li>Uses the distribution of QC metrics to set thresholds</li>
<li>More robust to outliers</li>
<li>Can be automated</li>
<li><em>MAD = median(|x - median(x)|)</em>, where <em>x</em> is the QC metric</li>
</ul></li>
</ul>
</section>
<section id="quality-control-5" class="slide level2 smaller">
<h2>Quality Control</h2>
<ul>
<li>Best to be as permissive as possible
<ul>
<li>Don’t want to filter out real cells</li>
</ul></li>
<li>Can always filter more later</li>
<li>Iterative and data driven approach is best
<ul>
<li>But beware of <a href="https://sites.stat.columbia.edu/gelman/research/unpublished/p_hacking.pdf">The garden of forking paths</a>
<ul>
<li>Why multiple comparisons can be a problem, even when there is no “fishing expedition” or “p-hacking”</li>
<li>Be careful with your conclusions</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="doublet-detection" class="slide level2">
<h2>Doublet Detection</h2>
<ul>
<li>Can be filtered out using tools like the <code>scDblFinder</code> R package or <code>Scrublet</code> Python package</li>
</ul>
<p><img data-src="img/doublet-biorender.png" class="" width="200"></p>
</section>
<section id="correction-of-ambient-rna" class="slide level2 smaller">
<h2>Correction of ambient RNA</h2>
<ul>
<li>RNA in the environment of the cell
<ul>
<li>Can be from dead cells, lysed cells, or other sources</li>
</ul></li>
<li>Potentially a significant source of noise in single cell RNAseq data</li>
<li>Can be corrected for using tools like <code>SoupX</code>
<ul>
<li>These tools use the expression profiles of empty droplets to estimate the ambient RNA profile</li>
<li>They then subtract this profile from the expression data of each cell</li>
</ul></li>
</ul>

<img data-src="img/droplet_ambient-biorender.png" width="200" class="r-stretch"></section>
<section id="normalization" class="slide level2 smaller">
<h2>Normalization</h2>
<ul>
<li>Each step of single cell workflow introduces a degree of variablity
<ul>
<li>Capture of cells and mRNA molecules</li>
<li>Reverse transcription</li>
<li>Amplification</li>
<li>Sequencing</li>
</ul></li>
<li>Count matrix contains widely varying variance terms</li>
<li>Statistical methods assume uniform variance</li>
</ul>
</section>
<section id="normalization---depth-bias" class="slide level2 smaller">
<h2>Normalization - Depth bias</h2>
<ul>
<li>Two genes: A &amp; B</li>
<li>Two cell types: Blue &amp; Green</li>
<li>Normalize by dividing UMI counts for each gene by total</li>
</ul>

<img data-src="img/depth_bias-cambridge_institute.png" class="r-stretch quarto-figure-center"><p class="caption">Cambridge Institute</p><ul>
<li>There is not differential expression, only a difference in sequencing depth</li>
</ul>
</section>
<section id="normalization-1" class="slide level2 smaller">
<h2>Normalization</h2>
<ul>
<li>Normalization adjusts the raw counts by scaling to a speciied range.</li>
<li>Reduces technical differences so that differences between are primarily biological</li>
<li>See <a href="https://doi.org/10.1038/s41592-023-01814-1">Ahlmann-Eltze et al.&nbsp;2020</a> for review &amp; benchmarking of normalization methods</li>
<li>Different techniques are better suited to different downstream analyses</li>
</ul>
</section>
<section id="normalization---techniques" class="slide level2 smaller">
<h2>Normalization - Techniques</h2>
<ul>
<li>Shifted logarithm
<ul>
<li>Works well for dimensionality reduction and differential expression</li>
</ul></li>
<li>Scran’s pooling-based size factor estimation method
<ul>
<li>Works well or batch correction<br>
</li>
</ul></li>
<li>Analytic Pearson residuals</li>
</ul>
</section>
<section id="feature-selection" class="slide level2 smaller">
<h2>Feature Selection</h2>
<ul>
<li>Many genes are not informative for downstream analysis</li>
<li>We want to:
<ul>
<li>Select gene that captue biologically meaningful variation</li>
<li>Reduce the number of genes that only contribute noise</li>
</ul></li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/feature_selection-luecken_theis_2019.jpeg" class="" width="800"></p>
<figcaption>Luecken &amp; Theis, 2019</figcaption>
</figure>
</div>
</section>
<section id="dimensionality-reduction" class="slide level2 smaller">
<h2>Dimensionality Reduction</h2>
<ul>
<li>scRNAseq data suffers from the “curse of dimensionality”
<ul>
<li>The data have a high number of dimensions (genes)</li>
<li>Data contains more noise and redundancy
<ul>
<li>Added dimensions do not add more information</li>
</ul></li>
</ul></li>
</ul>
</section>
<section id="dimensionality-reduction-1" class="slide level2 smaller">
<h2>Dimensionality Reduction</h2>
<ul>
<li>We already reduced the dimensionality of the data by selecting a subset of genes</li>
<li>We can reduce it further with dimensionality reduction algorithms</li>
</ul>

<img data-src="img/dimensionality_reduction-luecken_theis_2019.jpeg" class="r-stretch quarto-figure-center"><p class="caption">Luecken &amp; Theis, 2019</p><ul>
<li>The most popular:
<ul>
<li>Principal Component Analysis (PCA)</li>
<li>t-distributed Stochastic Neighbor Embedding (t-SNE)</li>
<li>Uniform Manifold Approximation and Projection (UMAP)</li>
</ul></li>
</ul>
</section>
<section id="principal-component-analysis-pca" class="slide level2 smaller">
<h2>Principal Component Analysis (PCA)</h2>
<ul>
<li>PCA creates new set of uncorrelated variables (principal components) that capture the most variance in the data</li>
<li>The first principal component captures the most variance, the second captures the second most, and so on</li>
<li>We can select the top principal components to use for downstream analysis</li>
<li>Highly efficient and easy to interpret</li>
<li>Not ideal for visualization</li>
</ul>
</section>
<section id="t-distributed-stochastic-neighbor-embedding-t-sne" class="slide level2 smaller">
<h2>t-distributed Stochastic Neighbor Embedding (t-SNE)</h2>
<ul>
<li>Graph based, non-linear dimensionality reduction technique</li>
<li>Only local distances are preserved</li>
<li>Stochastic, so results can vary between runs</li>
<li>Good for visualizing clusters in high-dimensional data</li>
<li>See <a href="https://youtu.be/NEaUSP4YerM">StatQuest Video</a> for a great explanation</li>
</ul>
</section>
<section id="uniform-manifold-approximation-and-projection-umap" class="slide level2 smaller">
<h2>Uniform Manifold Approximation and Projection (UMAP)</h2>
<ul>
<li>Graph based, non-linear dimensionality reduction technique</li>
<li>Similar in principle to t-SNE</li>
<li>Uses principal components to create a graph</li>
<li>Preserves both local and global structure</li>
</ul>
</section>
<section id="integration" class="slide level2 smaller">
<h2>Integration</h2>
<ul>
<li>Batch effects are a major challenge in scRNA-seq data analysis
<ul>
<li>Arise from processing cells in separate batches
<ul>
<li>Such as individual sample</li>
</ul></li>
<li>Obscure biological realities</li>
</ul></li>
<li>Removing batch effects is essential for analyses utlizing multiple batches</li>
</ul>

<img data-src="img/batch_effect-10x.png" width="700" class="r-stretch quarto-figure-center"><p class="caption">10X Genomics</p></section>
<section id="integration-1" class="slide level2 smaller">
<h2>Integration</h2>

<img data-src="img/integration-luecken_theis_2019.jpeg" width="750" class="r-stretch quarto-figure-center"><p class="caption">Luecken &amp; Theis, 2019</p></section>
<section id="clustering" class="slide level2">
<h2>Clustering</h2>
<ul>
<li>Clustering is the process of grouping cells based on their expression profiles</li>
<li>Common clustering algorithms:
<ul>
<li>K-means</li>
<li>Hierarchical clustering</li>
<li>Louvain clustering</li>
</ul></li>
<li>Louvain clustering is the most popular for scRNA-seq data
<ul>
<li>Can be used with UMAP or t-SNE for visualization</li>
</ul></li>
</ul>
</section>
<section id="clustering-1" class="slide level2">
<h2>Clustering</h2>
<ul>
<li>Keep in mind:
<ul>
<li>Clustering is an approximation of the underlying biological reality</li>
<li>Different levels can be appropriate for different questions
<ul>
<li>There may not be a single “correct” clustering</li>
</ul></li>
<li>Clustering algorithm will create as many cluster as you ask it to</li>
<li>Don’t overlook continuous variation</li>
</ul></li>
</ul>
</section>
<section id="annotation" class="slide level2 smaller">
<h2>Annotation</h2>
<ul>
<li>Assigning cell types to clusters</li>
<li>Can be done using:
<ul>
<li>Marker genes</li>
<li>Reference datasets</li>
<li>Automated annotation tools</li>
</ul></li>
<li>Marker genes are genes that are specifically expressed in a particular cell type</li>
<li>Reference datasets are datasets that have already been annotated</li>
</ul>
</section>
<section id="steps-2" class="slide level2">
<h2>Steps</h2>
<ol type="1">
<li>Quality control</li>
<li>Normalization</li>
<li>Dimensionality reduction</li>
<li>Clustering</li>
<li>Integration</li>
<li>Annotation</li>
</ol>

</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="06_analysis_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="06_analysis_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="06_analysis_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="06_analysis_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>