<!DOCTYPE html>
<html lang="en"><head>
<script src="01_introduction_files/libs/clipboard/clipboard.min.js"></script>
<script src="01_introduction_files/libs/quarto-html/tabby.min.js"></script>
<script src="01_introduction_files/libs/quarto-html/popper.min.js"></script>
<script src="01_introduction_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="01_introduction_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01_introduction_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="01_introduction_files/libs/quarto-html/quarto-syntax-highlighting-226bd0f977fa82dfae4534cac220d79a.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.23">

  <meta name="author" content="Kerry Cobb">
  <title>Single Cell Transcriptomics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="01_introduction_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="01_introduction_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
  </style>
  <link rel="stylesheet" href="01_introduction_files/libs/revealjs/dist/theme/quarto-250687541b6fd0303e6fa9a425de9713.css">
  <link href="01_introduction_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="01_introduction_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="01_introduction_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="01_introduction_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link href="custom.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Single Cell Transcriptomics</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Kerry Cobb 
</div>
</div>
</div>

</section>
<section id="objectives-learning-goals-for-workshop" class="slide level2 smaller">
<h2>Objectives &amp; Learning Goals for Workshop</h2>
<ul>
<li>HPC Usage</li>
<li>Singe Cell Transcriptome Concepts</li>
<li>Technical Details</li>
</ul>
</section>
<section id="objectives-learning-goals-for-workshop-1" class="slide level2 smaller">
<h2>Objectives &amp; Learning Goals for Workshop</h2>
<ul>
<li>HPC Usage
<ul>
<li>Accessing remote HPC</li>
<li>Submitting jobs</li>
<li>Basic linux usage</li>
</ul></li>
</ul>
</section>
<section id="objectives-learning-goals-for-workshop-2" class="slide level2 smaller">
<h2>Objectives &amp; Learning Goals for Workshop</h2>
<ul>
<li>Singe Cell Transcriptome Concepts
<ul>
<li>Sample preparation</li>
<li>Read mapping</li>
<li>Quality control</li>
<li>Analysis</li>
</ul></li>
</ul>
</section>
<section id="objectives-learning-goals-for-workshop-3" class="slide level2 smaller">
<h2>Objectives &amp; Learning Goals for Workshop</h2>
<ul>
<li>Technical Details
<ul>
<li>Available software</li>
<li>Scripts to run analyses</li>
</ul></li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="objectives-learning-goals-for-today" class="slide level2">
<h2>Objectives &amp; Learning Goals for today</h2>
<ul>
<li>Introduction to single cell transcriptomics</li>
<li>Overview of sample preparation, library construction, and sequencing</li>
<li>Initial data processing</li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="single-cell-transcriptomics" class="slide level2">
<h2>Single Cell Transcriptomics</h2>
<ul>
<li>Why do single cell transcriptomics?
<ul>
<li>Characterize heterogeneity among cells within cell population</li>
<li>Identify rare cell types</li>
<li>Explore interaction and communication among cells</li>
<li>Trace cell lineages during development</li>
</ul></li>
</ul>
</section>
<section id="bulk-vs-single-cell" class="slide level2 smaller">
<h2>Bulk vs Single Cell</h2>
<ul>
<li>Bulk transcriptomics
<ul>
<li>Measures average expression across all cells in a sample</li>
<li>Cannot detect heterogeneity among cells</li>
<li>Higher gene coverage</li>
</ul></li>
<li>Single cell transcriptomics
<ul>
<li>Measures expression in individual cells</li>
<li>Can detect heterogeneity among cells</li>
<li>Lower gene coverage</li>
</ul></li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="single-cell-transcriptomics-workflow" class="slide level2 smaller">
<h2>Single Cell Transcriptomics Workflow</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Tissue preparation</li>
</ol>
<p><img data-src="img/dissociation-biorender.png" data-top="0" height="150"></p>
<div class="fragment" data-fragment-index="1">
<ol start="2" type="1">
<li>Single cell isolation</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/cell-biorender.png" class="quarto-figure quarto-figure-center" height="75"></p>
</figure>
</div>
</div>
<div class="fragment" data-fragment-index="2">
<ol start="3" type="1">
<li>Library preparation</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/library_construct_3'-10x_genomics.png" class="quarto-figure quarto-figure-center" width="200"></p>
</figure>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment" data-fragment-index="3">
<ol start="4" type="1">
<li>Sequencing</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/novoseq.png" class="quarto-figure quarto-figure-center" height="200"></p>
</figure>
</div>
</div>
<div class="fragment" data-fragment-index="4">
<ol start="5" type="1">
<li>Data processing &amp; analysis</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/matrix.png" class="quarto-figure quarto-figure-center" height="200"></p>
</figure>
</div>
</div>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="tissue-preparation" class="slide level2 smaller">
<h2>1. Tissue Preparation</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="img/dissociation-biorender.png" height="150"></p>
<ul>
<li><span class="fragment highlight-red">Critical Step (Garbage in, garbage out)!</span></li>
<li>Many different methods
<ul>
<li>Details beyond the scope of this workshop</li>
</ul></li>
<li>Cell selction / enrichment</li>
<li>Sample clean-up</li>
</ul>
</div><div class="column" style="width:50%;">
<ul>
<li>Considerations for each method w/ regards to analysis
<ul>
<li>Expected cells</li>
<li>Quality of cells</li>
<li>Batch effects</li>
</ul></li>
</ul>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="tissue-preparation-1" class="slide level2 smaller">
<h2>1. Tissue Preparation</h2>
<ul>
<li>Take steps to remove batch effects &amp; confounding variables</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/confounding-baran-gale_et_al_2017.png" class="" width="600"></p>
<figcaption>Baran-Gale et al.&nbsp;2017</figcaption>
</figure>
</div>
<ul>
<li>Consider replication if feasible</li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="single-cell-isolation" class="slide level2 smaller">
<h2>2. Single Cell Isolation</h2>

<!------------------------------------------------------------------>
<img data-src="img/methods-svensson_2017.png" class="r-stretch quarto-figure-center"><p class="caption">Svensson et al.&nbsp;2017</p></section>
<section id="single-cell-isolation-1" class="slide level2 smaller">
<h2>2. Single Cell Isolation</h2>

<img data-src="img/isolation_workflow.png" class="r-stretch"><p>Multiple methods broadly categorized:</p>
<ol type="1">
<li>Emulsion based</li>
<li>Plate based</li>
</ol>
<aside class="notes">
<p>Presenter notes here.</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<!------------------------------------------------------------------>
</section>
<section id="section" class="slide level2 smaller">
<h2></h2>
<h3 id="single-cell-isolation---trade-off-between-cells-and-genes">Single Cell Isolation - Trade off between cells and genes</h3>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li>Depends on:
<ul>
<li>Method used</li>
<li>Choices made in implementation of method</li>
</ul></li>
<li>Required number of cells increases with complexity of sample</li>
<li>Sample can often be resequenced</li>
</ul>
</div><div class="column" style="width:40%;">
<p><img data-src="img/gene_cell_tradeoff.png"></p>
</div></div>
<aside class="notes">
<p>Not all of the library will be used up in most cases</p>
<style type="text/css">
        span.MJX_Assistive_MathML {
          position:absolute!important;
          clip: rect(1px, 1px, 1px, 1px);
          padding: 1px 0 0 0!important;
          border: 0!important;
          height: 1px!important;
          width: 1px!important;
          overflow: hidden!important;
          display:block!important;
      }</style></aside>
<!------------------------------------------------------------------>
</section>
<section id="single-cell-isolation---10x-genomics" class="slide level2">
<h2>Single Cell Isolation - 10X Genomics</h2>
<ul>
<li>By far the most popular approach</li>
<li>Source of all workshop data</li>
</ul>

<!------------------------------------------------------------------>
<img data-src="img/chromium-10x_genomics.webp" class="quarto-figure quarto-figure-center r-stretch" width="100"></section>
<section id="single-cell-isolation---10x-genomics-1" class="slide level2">
<h2>Single Cell Isolation - 10X Genomics</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/10x-10x_genomics.gif" class=""></p>
<figcaption>10X Genomics</figcaption>
</figure>
</div>
<!------------------------------------------------------------------>
</section>
<section id="section-1" class="slide level2 smaller">
<h2></h2>
<h3 id="single-cell-isolation---illumina-single-cell-3-rna">Single Cell Isolation - Illumina single cell 3’ RNA</h3>
<ul>
<li>Very recent offering</li>
<li>Formerly known as PIPSeq</li>
<li>Looks very promising, several UConn &amp; UCHC labs planning to adopt</li>
</ul>

<!------------------------------------------------------------------>
<img data-src="img/pipseq-clark_et_al_2023.png" class="r-stretch quarto-figure-center"><p class="caption">Clark et al.&nbsp;2023</p></section>
<section id="section-2" class="slide level2">
<h2></h2>
<h3 id="single-cell-isolation---combinatorial-barcoding">Single Cell Isolation - Combinatorial Barcoding</h3>
<h4 class="center" id="parse-biosciences-evercode">Parse Biosciences Evercode</h4>
<ul>
<li>Cells serve as reacion vessel</li>
<li>Serial splitting, barcoding, and pooling results in each cell having a unique combination of barcodes</li>
<li>C(96,4) = 3,321,960</li>
</ul>

<!------------------------------------------------------------------>
<img data-src="img/evercode-tran_et_al_2022.png" class="r-stretch quarto-figure-center"><p class="caption">Tran et al.&nbsp;2022</p></section>
<section id="section-3" class="slide level2">
<h2></h2>
<h3 id="single-cell-isolation---combinatorial-barcoding-1">Single Cell Isolation - Combinatorial Barcoding</h3>
<h4 class="center" id="scale-biosciences-quantum-barcoding">Scale Biosciences Quantum Barcoding</h4>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/quantum_barcoding-scale_biosciences.png" class=""></p>
<figcaption>Scale Biosciences</figcaption>
</figure>
</div>
<!------------------------------------------------------------------>
</section>
<section id="section-4" class="slide level2 smaller">
<h2></h2>
<h3 id="single-cell-isolation---smart-seq">Single Cell Isolation - SMART-SEQ</h3>
<ul>
<li>Plate-based</li>
<li>Much greater cost per cell</li>
<li>Can assay full transcripts</li>
</ul>

<!------------------------------------------------------------------>
<img data-src="img/smartseq-macosko_2020.png" class="r-stretch quarto-figure-center"><p class="caption">Macosko 2020</p></section>
<section id="single-cell-isolation---others" class="slide level2">
<h2>Single Cell Isolation - Others</h2>
<ul>
<li>Honeycomb HIVE</li>
<li>Singleron</li>
<li>Asteria</li>
<li>Fluent</li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="library-preparation" class="slide level2 smaller">
<h2>3. Library Preparation</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ol type="1">
<li>Convert RNA to cDNA
<ul>
<li>Reverse transcription</li>
</ul></li>
<li>Fragmentation</li>
<li>Barcode ligation
<ul>
<li>Unique molecular identifiers (UMIs)</li>
<li>Cell barcodes</li>
</ul></li>
<li>Sequencing adapter ligation</li>
<li>Amplification</li>
<li>Remove contimating RNA and cell debris</li>
</ol>
</div><div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/flow_cell_oligos-Illumina.png" class="" width="500"></p>
<figcaption>Illumina</figcaption>
</figure>
</div>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="x-library-preparation" class="slide level2 center">
<h2>10X Library Preparation</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/droplet_bead.png" class="nostretch quarto-figure quarto-figure-center" width="200"></p>
</figure>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/droplet_bead_lysed.png" class="nostretch quarto-figure quarto-figure-center" width="200"></p>
</figure>
</div>
</div>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="x-rna-capture" class="slide level2">
<h2>10X RNA Capture</h2>

<!------------------------------------------------------------------>
<img data-src="img/capture-adapted-10x_genomics.png" class="r-stretch quarto-figure-center"><p class="caption">Adapted from 10X Genomics</p></section>
<section id="x-3-vs-5-libraries" class="slide level2">
<h2>10X 3` vs 5` libraries</h2>
<p><img data-src="img/bead_capture_3'-10x_genomics.png" width="500"></p>
<p><img data-src="img/bead_capture_5'-10x_genomics.png" width="500"></p>
<ul>
<li>Refers to the end of an RNA molecule that is sequenced</li>
<li>3` came first</li>
<li>5` permits identification of T-cell and B-cell receptor sequence</li>
<li>Sensitivity is comparable according to 10X</li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="x-3-capture" class="slide level2">
<h2>10X 3` Capture</h2>

<!------------------------------------------------------------------>
<img data-src="img/capture_3'-10x_genomics.png" class="r-stretch quarto-figure-center"><p class="caption">10X Genomics</p></section>
<section id="x-5-capture" class="slide level2">
<h2>10X 5` Capture</h2>

<!------------------------------------------------------------------>
<img data-src="img/capture_5'-10x_genomics.png" class="r-stretch quarto-figure-center"><p class="caption">10X Genomics</p></section>
<section id="x-3-library-construction" class="slide level2">
<h2>10X 3’ Library construction</h2>

<!------------------------------------------------------------------>
<img data-src="img/library_construction_3'-10x_genomics.png" class="r-stretch quarto-figure-center"><p class="caption">10X Genomics</p></section>
<section id="x-5-library-construction" class="slide level2">
<h2>10X 5’ Library construction</h2>

<!------------------------------------------------------------------>
<img data-src="img/library_construction_5'-10x_genomics.png" class="r-stretch quarto-figure-center"><p class="caption">10X Genomics</p></section>
<section id="constructing-libraries-from-mrna" class="slide level2">
<h2>Constructing Libraries from mRNA</h2>
<ul>
<li>10X is most common
<ul>
<li>Multiple versions</li>
</ul></li>
<li>Other methods exist</li>
<li>Method used can have important implications for analysis
<ul>
<li>Index hopping</li>
<li>PCR duplicates</li>
<li>Biased amplification</li>
<li>Biased capture?</li>
</ul></li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="unique-molecular-identifiers-umis" class="slide level2">
<h2>Unique Molecular Identifiers (UMIs)</h2>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/umi-adapted-10x_genomics.png" class="" width="700"></p>
<figcaption>Adapted from 10X Genomics</figcaption>
</figure>
</div>
<ul>
<li>Permit identification of PCR duplicates</li>
<li>Results in more accurate estimate of expression</li>
<li>10X Genomics uses a random 12 nt sequence on each oligo</li>
</ul>
<!------------------------------------------------------------------>
</section>
<section id="sequencing" class="slide level2">
<h2>4. Sequencing</h2>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li>Typically done with Illumina short reads
<ul>
<li>Focus of workshop</li>
</ul></li>
<li>Could be done with any sequencing platform
<ul>
<li>Each may require special considerations
<ul>
<li>Read length</li>
<li>Error rate</li>
<li>Technical artifacts/biases</li>
</ul></li>
</ul></li>
</ul>
</div><div class="column" style="width:40%;">
<p><img data-src="img/novoseq.png" height="300"></p>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="multiplexing-samples" class="slide level2 smaller">
<h2>Multiplexing samples</h2>
<div class="columns">
<div class="column" style="width:60%;">
<ul>
<li>Samples can be multiplexed</li>
<li>Two approaches
<ol type="1">
<li>Cell labelling</li>
<li>SNP variation</li>
</ol></li>
<li>Can reduce impact of batch effects</li>
<li>Cell labelling permits overloading
<ul>
<li>Reduces empty droplets, doublets easy to identify</li>
</ul></li>
</ul>
</div><div class="column" style="width:40%;">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="img/cell_hashing-stoeckius&amp;smibert_2018.png"></p>
<figcaption>Stoeckius &amp; Smibert 2018</figcaption>
</figure>
</div>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="probe-capture" class="slide level2 smaller">
<h2>Probe capture</h2>
<ul>
<li>Generally</li>
<li>RNA captured by hybridization with probes rather than poly A tail</li>
<li>Permits sequencing of formalin fixed tissue and low quality tissues</li>
<li>Adds additional cost</li>
<li>Limited to humans and mice</li>
</ul>

<!------------------------------------------------------------------>
<img data-src="img/probe_capture-10x_genomics.png" class="r-stretch quarto-figure-center"><p class="caption">10X Genomics</p></section>
<section id="single-cell-vs-single-nuclei" class="slide level2 smaller">
<h2>Single-cell vs Single-nuclei</h2>
<div class="columns">
<div class="column" style="width:45%;">
<h3 id="single-cell">Single Cell</h3>
<p><img data-src="img/cell.png" class="nostretch" width="200"></p>
<ul>
<li>Can detect transcripts in the cytoplasm as well as the nucleus</li>
<li>Typically want to use fresh cells</li>
<li>Signal more prone to perturbation caused by tissue processing</li>
</ul>
</div><div class="column" style="width:10%;">

</div><div class="column" style="width:45%;">
<h3 id="single-nuclei">Single Nuclei</h3>
<p><img data-src="img/nucleus.png" class="nostretch" width="200"></p>
<ul>
<li>Can detect more unprocessed mRNA containing introns</li>
<li>Cannot detect transcripts in the cytoplasm making it unsuitable for some investigations</li>
<li>Can be use with preserved cells and difficult to dissociate cells</li>
</ul>
</div></div>
<!------------------------------------------------------------------>
</section>
<section id="data-processing" class="slide level2">
<h2>5. Data Processing</h2>
<ul>
<li>Up next!</li>
</ul>

</section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="01_introduction_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="01_introduction_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="01_introduction_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="01_introduction_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>